python/ray/tune/trial_executor.py

```diff
@@ -15,16 +15,55 @@ class TrialExecutor(object):
     and starting/stopping trials.
     """
 
-    def __init__(self, queue_trials=False):
+    def __init__(self, queue_trials=False, checkpoint_mode=False):
         """Initializes a new TrialExecutor.
 
         Args:
             queue_trials (bool): Whether to queue trials when the cluster does
                 not currently have enough resources to launch one. This should
                 be set to True when running on an autoscaling cluster to enable
                 automatic scale-up.
+            checkpoint_mode (bool): Whether to track metadata on status
+                change.
         """
         self._queue_trials = queue_trials
+        self._checkpoint_mode = checkpoint_mode
+        self._checkpoints = {}
+
+    def set_status(self, trial, status):
+        """Sets status and checkpoints metadata if needed.
+
+        Only checkpoints metadata if trial status is a terminal condition.
+        PENDING, PAUSED, and RUNNING switches have checkpoints taken care of
+        in the TrialRunner.
+
+        Args:
+            trial (Trial): Trial to checkpoint.
+            status (Trial.status): Status to set trial to.
+        """
+        trial.status = status
+        if status in [Trial.TERMINATED, Trial.ERROR]:
+            self.try_checkpoint_metadata(trial)
+
+    def try_checkpoint_metadata(self, trial):
+        """Checkpoints metadata if checkpoint_mode is True.
+
+        Args:
+            trial (Trial): Trial to checkpoint.
+        """
+        if self._checkpoint_mode:
+            if trial._checkpoint.storage == Checkpoint.MEMORY:
+                logger.debug("Not saving data for trial w/ memory checkpoint.")
+                return
```

Comment:
Won't this make these trials disappear on resume? Why not just ignore the memory checkpoint instead of dropping the trial data entirely?

Reply:
No; the trial state cache isn't cleared at this point -- it's only skipping an update to the cache.`checkpoint_freq` always checkpoints to disk; only PBT checkpoints to memory. If a trial has checkpointing turned on, it'll have some disk checkpoint in cache.